src/id/shinigami/src/eu/kanade/tachiyomi/extension/id/shinigami/Shinigami.kt

```diff
@@ -83,70 +56,195 @@ class Shinigami : Madara("Shinigami", "https://shinigami09.com", "id"), Configur
         .rateLimit(3)
         .build()
 
-    // Tags are useless as they are just SEO keywords.
-    override val mangaDetailsSelectorTag = ""
+    override fun headersBuilder(): Headers.Builder = Headers.Builder()
+        .add("X-Requested-With", randomString((1..20).random())) // added for webview, and removed in interceptor for normal use
 
-    override val chapterUrlSelector = "div.chapter-link:not([style~=display:\\snone]) a"
+    private fun randomString(length: Int): String {
+        val charPool = ('a'..'z') + ('A'..'Z')
+        return List(length) { charPool.random() }.joinToString("")
+    }
+
+    private fun apiHeadersBuilder(): Headers.Builder = headersBuilder()
+        .add("Accept", "application/json")
+        .add("DNT", "1")
+        .add("Origin", baseUrl)
+        .add("Sec-GPC", "1")
+
+    override fun popularMangaRequest(page: Int): Request {
+        val url = "$apiUrl/v1/manga/list".toHttpUrl().newBuilder()
+            .addQueryParameter("page", page.toString())
+            .addQueryParameter("page_size", "30")
+            .addQueryParameter("sort", "popularity")
+            .build()
 
-    override fun chapterFromElement(element: Element): SChapter = SChapter.create().apply {
-        val urlElement = element.selectFirst(chapterUrlSelector)!!
+        return GET(url, apiHeaders)
+    }
+
+    override fun popularMangaParse(response: Response): MangasPage {
+        val rootObject = response.parseAs<ShinigamiBrowseDto>()
+        val projectList = rootObject.data.map(::popularMangaFromObject)
 
-        name = urlElement.selectFirst("p.chapter-manhwa-title")?.text()
-            ?: urlElement.ownText()
-        date_upload = urlElement.selectFirst("span.chapter-release-date > i")?.text()
-            .let { parseChapterDate(it) }
+        val hasNextPage = rootObject.meta.page < rootObject.meta.totalPage
 
-        val fixedUrl = urlElement.attr("abs:href")
+        return MangasPage(projectList, hasNextPage)
+    }
 
-        setUrlWithoutDomain(fixedUrl)
+    private fun popularMangaFromObject(obj: ShinigamiBrowseDataDto): SManga = SManga.create().apply {
+        title = obj.title.toString()
+        thumbnail_url = obj.thumbnail
+        url = obj.mangaId.toString()
```

Comment:
`.toString()` hides that the value might be null.
```suggestion
        title = obj.title!!
        thumbnail_url = obj.thumbnail
        url = obj.mangaId!!
```